src/util/ringdelaybuffer.cpp

```diff
@@ -0,0 +1,215 @@
+/*
+* $Id$
+* Portable Audio I/O Library
+* Ring Buffer utility.
+*
+* Author: Phil Burk, http://www.softsynth.com
+* modified for SMP safety on OS X by Bjorn Roche.
+* also allowed for const where possible.
+* modified for multiple-byte-sized data elements by Sven Fischer
+*
+* Note that this is safe only for a single-thread reader
+* and a single-thread writer.
+*
+* This program is distributed with the PortAudio Portable Audio Library.
+* For more information see: http://www.portaudio.com
+* Copyright (c) 1999-2000 Ross Bencina and Phil Burk
+*
+* Permission is hereby granted, free of charge, to any person obtaining
+* a copy of this software and associated documentation files
+* (the "Software"), to deal in the Software without restriction,
+* including without limitation the rights to use, copy, modify, merge,
+* publish, distribute, sublicense, and/or sell copies of the Software,
+* and to permit persons to whom the Software is furnished to do so,
+* subject to the following conditions:
+*
+* The above copyright notice and this permission notice shall be
+* included in all copies or substantial portions of the Software.
+*
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR
+* ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
+* CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+* WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+/*
+* The text above constitutes the entire PortAudio license; however,
+* the PortAudio community also makes the following non-binding requests:
+*
+* Any person wishing to distribute modifications to the Software is
+* requested to send the modifications to the original developer so that
+* they can be incorporated into the canonical version. It is also
+* requested that these non-binding requests be included along with the
+* license above.
+*/
+
+#include "ringdelaybuffer.h"
+
+RingDelayBuffer::RingDelayBuffer(SINT bufferSize)
+        : m_bufferSize(bufferSize),
+          m_readPos(0),
+          m_writePos(0),
+          m_ringFullMask((bufferSize * 2) - 1),
+          m_ringMask(bufferSize - 1),
+          m_jumpLeftAroundMask(0),
+          m_buffer(bufferSize) {
+    // TODO(davidchocholaty) Handle to allow only power of two for the size of the ring delay buffer.
+
+    // Set the ring delay buffer items to 0.
+    CSAMPLE* bufferData = m_buffer.data();
+    memset(bufferData, 0, sizeof(*bufferData) * bufferSize);
+}
+
+SINT RingDelayBuffer::read(CSAMPLE* pBuffer, SINT itemsToRead) {
+    const SINT available = getReadAvailable();
+
+    VERIFY_OR_DEBUG_ASSERT(itemsToRead <= available) {
+        itemsToRead = available;
+    }
+
+    const SINT position = m_readPos & m_ringMask;
```

Comment:
I am not sure if the concept of a member variable m_readPos is still suitable. 

For my understanding we have the current sample at m_writePos - 1 and the delayed samples before that so the actual read position is m_writePos - "current delay" - itemsToRead  

We need to make sure this does not hit the end of the buffer. 

Idea: Pass the delay along to the read function?

Reply:
Thank you so much for this idea. Sounds cool and I think it cleans up the code a lot. I have just one thing on my mind for two cases. Based on the proposed new version, the ring delay buffer will behave a little bit differently.

1. If, for example, 8 samples will be written into the ring delay buffer, and then the different amount of samples will be read with zero delays, just propose 4 samples, so, for the previous version, the ring buffer will read the first half of the written samples by index 0, however, the new version will read the second half of the written samples (8 - 0 - 4).

2. If the write method will be called two times in a row without calling the read method, the situation will be quite similar to the first one.

For our future use case for `EngineEffectsDelay`, this cannot happen. Anyway, if I will think only about `RingDelayBuffer`, it may be possible. What do you think about that? I don't want to throw down somehow this idea, I really like it and I would like to use it. I'm only thinking about possible cases, which can happen.

Reply:
I would like to re-open a discussion about this idea. I would like to discuss just one thing which I noticed. If the new version will be used, then a lot of checks of delay value have to be run for every `read()` call, probably all code in `moveReadPositionBy()`. Can I ask what is your view on this situation?